### PR TITLE
ENHANCE: Make groups lockable

### PR DESCRIPTION
**Pull request by jdamatopoulos**
# Description

This PR implements the front-end of the feature "lockable-groups". This allows groups & group sets to be locked by staff members, which then restricts students from making changes (i.e. the title/the members by leaving/joining) to locked groups.

A short showcase video of the feature can be seen here: https://youtu.be/hzRNAqd3NCs

The corresponding back-end PR for this feature can be found here: https://github.com/doubtfire-lms/doubtfire-api/pull/284

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- Created a new student group (unlocked upon creation) and verified that students are allowed to change its title, leave it, and join other locked groups.
- Locked the group and verified that the aforementioned changes are prevented.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/app/units/states/edit/directives/unit-group-set-editor/unit-group-set-editor.coffee [hunk 1]
```diff
@@ -40,6 +41,31 @@ angular.module('doubtfire.units.states.edit.directives.unit-group-set-editor', [
...
+          $scope.$broadcast 'UnitGroupSetEditor/SelectedGroupSetChanged', { id: gs.id }
+
+        (response) ->
+          alertService.add("danger", "Failed to #{if gs.locked then 'unlock' else 'lock'} #{gs.name}", 6000)
+      )
+
+    # Listen for groups all groups being locked.
+    listeners.push $scope.$on 'GroupSelector/GroupLockToggled', (e, args) ->
+      if args.groupSetId == $scope.selectedGroupSet.id
+        $scope.selectedGroupSet.locked = args.allLocked
```
Comment (macite) [inline src/app/units/states/edit/directives/unit-group-set-editor/unit-group-set-editor.coffee:67]: Should only be when true - as a group set may be locked when all groups are not locked.

Comment (RavinduL) [reply src/app/units/states/edit/directives/unit-group-set-editor/unit-group-set-editor.coffee]: Context about the changes that are requested:

> What I would like is to also have a lock on the group set
> 
> So each group is locked if it is locked OR the group set is locked
> 
> To lock all groups you then just lock the group set
> 
> Simplifies the check for all groups being locked - as you check just the group set
> 
> Also means you can have some individual groups locked, and then lock the group set, if you unlock the group set then those that were originally locked will remain locked

Comment (jdamatopoulos) [reply src/app/units/states/edit/directives/unit-group-set-editor/unit-group-set-editor.coffee]: We have made the changes in a new commit, which can be viewed here: https://github.com/doubtfire-lms/doubtfire-web/pull/281/commits/2d683e9ffc305b143aad3a991d3146fc35ce27e9.

Thank you!

Comment (RavinduL) [reply src/app/units/states/edit/directives/unit-group-set-editor/unit-group-set-editor.coffee]: The corresponding backend commit is: https://github.com/doubtfire-lms/doubtfire-api/pull/284/commits/7af266b4e18de04490c880e3790c712242ee0a4a

A video demonstrating the effect of this change is at https://youtu.be/faWO8D796Wo

**Pull request merged**